File: mjx/state.py
from __future__ import annotations

from typing import List, Optional, Tuple

import _mjx  # type: ignore
from google.protobuf import json_format

import mjxproto
from mjx.action import Action
from mjx.observation import Observation
from mjx.visualizer.svg import save_svg, show_svg, to_svg


class State:
    def __init__(self, state_json: Optional[str] = None) -> None:
        self._cpp_obj: Optional[_mjx.State] = None  # type: ignore
        if state_json is None:
            return

        self._cpp_obj = _mjx.State(state_json)  # type: ignore

    def _repr_html_(self) -> None:
        observation = self.to_proto()
        return to_svg(observation, target_idx=None)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, State):
            raise NotImplementedError
        raise NotImplementedError  # TODO: implement

    def __ne__(self, other: object) -> bool:
        if not isinstance(other, State):
            raise NotImplementedError
        raise NotImplementedError  # TODO: implement

    def to_json(self) -> str:
        assert self._cpp_obj is not None
        return self._cpp_obj.to_json()  # type: ignore

    def to_proto(self) -> mjxproto.State:
        assert self._cpp_obj is not None
        return json_format.Parse(self.to_json(), mjxproto.State())

    def past_decisions(self) -> List[Tuple[Observation, Action]]:
        return [
            (Observation._from_cpp_obj(obs), Action._from_cpp_obj(act))
            for obs, act in self._cpp_obj.past_decisions()
        ]

    def save_svg(self, filename: str, view_idx: int = 0):
        assert filename.endswith(".svg")
        assert 0 <= view_idx < 4

        observation = self.to_proto()
        save_svg(observation, filename, view_idx)

    def show_svg(self, view_idx: Optional[int] = None) -> None:
        assert view_idx is None or 0 <= view_idx < 4

        observation = self.to_proto()
        show_svg(observation, target_idx=view_idx)

    @classmethod
    def from_proto(cls, proto: mjxproto.State) -> State:
        return State(json_format.MessageToJson(proto))

    @classmethod
    def _from_cpp_obj(cls, cpp_obj) -> State:
        state = cls()
        state._cpp_obj = cpp_obj
        return state

File: include/mjx/state.cpp
#include "mjx/state.h"

#include <google/protobuf/util/json_util.h>
#include <google/protobuf/util/message_differencer.h>

#include <utility>

#include "mjx/internal/state.h"

namespace mjx {
mjx::State::State(mjxproto::State proto) : proto_(std::move(proto)) {}

State::State(const std::string& json) {
  auto status = google::protobuf::util::JsonStringToMessage(json, &proto_);
  assert(status.ok());
}

const mjxproto::State& mjx::State::proto() const noexcept { return proto_; }

std::string mjx::State::ToJson() const noexcept {
  std::string serialized;
  auto status =
      google::protobuf::util::MessageToJsonString(proto_, &serialized);
  assert(status.ok());
  return serialized;
}

bool State::operator==(const State& other) const noexcept {
  return google::protobuf::util::MessageDifferencer::Equals(proto_,
                                                            other.proto_);
}

bool State::operator!=(const State& other) const noexcept {
  return !(*this == other);
}

std::vector<std::pair<Observation, Action>> State::past_decisions()
    const noexcept {
  std::vector<std::pair<Observation, Action>> decisions;
  auto proto_decisions = internal::State::GeneratePastDecisions(proto());
  for (const auto& [obs, action] : proto_decisions) {
    decisions.emplace_back(Observation(obs), Action(action));
  }
  return decisions;
}
}  // namespace mjx
